src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
+  mName <-
+    -- May happen when used on data families
+    maybe (fail $ show $ mArgNotVar typeName mArg) return $ tvName' mArg
+
+  return CLInfo {
+    typeName, funName, conName, effTArgs, mName, resArg, conCxt, conArgs
+    }
+  where
+    mArgNotVar name mArg
+      =   text "Monad argument ‘" <> ppr mArg <> text "’ in effect ‘"
+          <> ppr name <> text "’ is not a type variable"
+
+    missingEffTArgs name
+      =   text "Effect ‘" <> ppr name
+          <> text "’ has not enough type arguments"
+      $+$ nest 4
+          (   text "At least monad and result argument are required, e.g.:"
+          $+$ nest 4
+              (   text ""
+              $+$ ppr (DataD [] base args Nothing [] []) <+> text "..."
+              $+$ text ""
+              )
+          )
+      where
+        base = mapNameBase id name
+        args = PlainTV . mkName <$> ["m", "a"]
+
+------------------------------------------------------------------------------
+-- | Generates signature for lifting function and type arguments to apply in
+-- it's body on effect's data constructor.
+genSig :: ConLiftInfo -> Q Dec
+genSig CLInfo {
+    typeName, funName, effTArgs, mName, resArg, conCxt, conArgs
+  } = do
+  r <- newName "r"
+  let memberCxt  = ConT ''Member `AppT` eff `AppT` VarT r
+      eff        = foldl' AppT (ConT typeName) effTArgs
+      -- TODO: can use of 'm' blow up somehow?
+      funArgs    = replaceMArg mName r conArgs
+      returnType = ConT ''Sem `AppT` VarT r `AppT` resArg
+
+  return $ SigD funName $ quantifyType $
+    ForallT [] (memberCxt : conCxt) $ foldArrows $ funArgs ++ [returnType]
+
+  where
+    -- Replaces use of @m@ in type with @Sem r@.
+    replaceMArg m r = applySubstitution
+                    $ M.singleton m (ConT ''Sem `AppT` VarT r)
+
+------------------------------------------------------------------------------
 -- | Builds a function definition of the form @x a b c = send $ X a b c@.
-genDecl :: Con -> Q [Dec]
-genDecl (ForallC _       _     con) = genDecl con
-genDecl (GadtC   [cName] tArgs _  ) = do
-  let fnName = getDeclName cName
-  let arity  = length tArgs - 1
-  dTypeVars <- forM [0 .. arity] $ const $ newName "a"
-  pure $
-    [PragmaD (InlineP fnName Inlinable ConLike AllPhases)
-    , FunD fnName . pure $ Clause
-        (VarP <$> dTypeVars)
-        (NormalB . AppE (VarE 'send) $ foldl
-          (\b -> AppE b . VarE)
-          (ConE cName)
-          dTypeVars
-        )
-        []
+genDec :: ConLiftInfo -> Q [Dec]
+-- TODO: apply effect's type arguments to avoid "bad ambiguity"
+genDec CLInfo { funName, conName, conArgs } = do
+  funArgsNames <- replicateM (length conArgs) $ newName "x"
+
+  return [
+      PragmaD $ InlineP funName Inlinable ConLike AllPhases
+    , FunD funName [
+        Clause (VarP <$> funArgsNames)
+               (NormalB $ AppE (VarE 'send) $
+                 foldl1 AppE $ ConE conName : fmap VarE funArgsNames)
```

Comment:
Maybe some helpers for this kind of thing would go a long way. I wouldn't be surprised if they exist in `TH.Syntax`

Reply:
Will look into it.

Reply:
Can't find any useful - should I create some helpers locally? Because I'm not sure if it's really worth it.

Reply:
It's pity that quotations aren't more flexible - I can't for example splice function name into declaration...

Reply:
Don't worry about it in that case. There are more important things in the world than pretty-looking TH.

Reply:
Ok